### PR TITLE
[SPARK-15390] fix broadcast with 100 millions rows

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

When broadcast a table with more than 100 millions rows (should not ideally), the size of needed memory will overflow.

This PR fix the overflow by converting it to Long when calculating the size of memory.

Also add more checking in broadcast to show reasonable messages.
## How was this patch tested?

Add test.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala [hunk 1]
```diff
@@ -72,9 +72,18 @@ case class BroadcastExchangeExec(
         val beforeCollect = System.nanoTime()
         // Note that we use .executeCollect() because we don't want to convert data to Scala types
         val input: Array[InternalRow] = child.executeCollect()
+        if (input.length >= (512 << 20)) {
+          throw new SparkException(
+            s"Cannot broadcast the table with more than 512 millions rows: ${input.length} rows")
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala]: this is technically not 512 million isn't it?

**Comment by rxin** [comment]
cc @sameeragarwal

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala (hunk 1 above)
Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala]: I think it'd be good to make these 2 consistent (either use 512 << 20 or 512000000 at both places)

**Comment by sameeragarwal** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala (hunk 1 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala]: Yes, it's not, will update them.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 2]
```diff
@@ -410,9 +410,10 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
 
   private def init(): Unit = {
     if (mm != null) {
+      require(capacity < (512 << 20), "Cannot broadcast more than 512 millions rows")
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Is `capacity` number of row?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: yes

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Looks like it is.

**Comment by davies** [comment]
Merging this into master and 2.0, thanks!

**Comment by mengxr** [comment]
@davies @rxin It seems that this PR caused OOO in master builds. 

~~~
*** RUN ABORTED ***
  java.lang.OutOfMemoryError: Java heap space
  at org.apache.spark.sql.execution.joins.LongToUnsafeRowMap.init(HashedRelation.scala:417)
  at org.apache.spark.sql.execution.joins.LongToUnsafeRowMap.<init>(HashedRelation.scala:423)
  at org.apache.spark.sql.execution.joins.LongHashedRelation$.apply(HashedRelation.scala:792)
  at org.apache.spark.sql.execution.joins.HashedRelationSuite$$anonfun$6.apply$mcV$sp(HashedRelationSuite.scala:227)
  at org.apache.spark.sql.execution.joins.HashedRelationSuite$$anonfun$6.apply(HashedRelationSuite.scala:216)
  at org.apache.spark.sql.execution.joins.HashedRelationSuite$$anonfun$6.apply(HashedRelationSuite.scala:216)
  at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
  at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
~~~

https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-maven-hadoop-2.2/1066/consoleFull

**Comment by davies** [comment]
@mengxr Disable this test in master and 2.0.

**Pull request closed without merging**